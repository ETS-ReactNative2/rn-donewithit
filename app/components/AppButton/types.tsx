import { GestureResponderEvent, StyleProp, ViewStyle } from 'react-native'

export type AppButtonProps = {
  title: string
  fontSize?: number
  height?: number
  backgroundColor?: string
  borderRadius?: number
  style?: StyleProp<ViewStyle>
  onPress?: (e?: GestureResponderEvent) => void
}

export type useStylesProps = {
  fontSize?: number
  height?: number
  backgroundColor?: string
  borderRadius?: number
}
