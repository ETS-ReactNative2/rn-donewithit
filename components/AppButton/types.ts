import { StyleProp, ViewStyle } from 'react-native'

export type AppButtonProps = {
  title?: string
  fontSize?: number
  height?: number
  backgroundColor?: string
  borderRadius?: number
  style?: StyleProp<ViewStyle>
  // onPress?: (e?: GestureResponderEvent) => void
  onPress?: (e?: any) => void // fix for FormEvent<HTMLFormElement> eventType.
}

export type useStylesProps = {
  fontSize?: number
  height?: number
  backgroundColor?: string
  borderRadius?: number
}
