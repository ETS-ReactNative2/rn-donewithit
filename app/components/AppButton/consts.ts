import COLORS from '../../config/colors'

const CONSTS = {
  BUTTON_BORDER_RADIUS: 25,
  BUTTON_HEIGHT: 50,
  BUTTON_TITLE: 'button',
  FONT_SIZE: 18,
  HEIGHT: 50,
  BACKGROUND_COLOR: COLORS.PRIMARY
}

export default CONSTS
