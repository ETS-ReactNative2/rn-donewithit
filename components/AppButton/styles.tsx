import { StyleSheet } from 'react-native';
import colors from '../../config/colors';

const styles = StyleSheet.create({
  buttonBox: {
    borderRadius: 25,
    overflow: 'hidden',
  },
  button: {
    height: 50,
    borderRadius: 25,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: colors.primary,
  },
  text: {
    fontSize: 18,
    fontWeight: 'bold',
    color: colors.white,
  }
})

export default styles
