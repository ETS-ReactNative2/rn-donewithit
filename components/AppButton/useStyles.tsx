import React, { useMemo } from 'react'
import { StyleSheet, Platform } from 'react-native'
import CONSTS from './consts'
import COLORS from '../../config/colors'
import { useStylesProps } from './types'

function useStyles({
  height = CONSTS.BUTTON_HEIGHT,
  fontSize = CONSTS.FONT_SIZE,
  backgroundColor = CONSTS.BACKGROUND_COLOR,
  borderRadius = CONSTS.BUTTON_BORDER_RADIUS
}: useStylesProps) {
  const styles = useMemo(
    () =>
      StyleSheet.create({
        buttonBox: {
          ...Platform.select({
            android: {
              borderRadius,
              overflow: 'hidden'
            }
          })
        },
        button: {
          height,
          borderRadius,
          justifyContent: 'center',
          alignItems: 'center',
          backgroundColor
        },
        text: {
          fontSize,
          fontWeight: 'bold',
          color: COLORS.WHITE,
          textTransform: 'uppercase'
        }
      }),
    []
  )

  return styles
}

export default useStyles
