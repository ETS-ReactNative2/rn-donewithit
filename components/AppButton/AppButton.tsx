import React from 'react'
import {
  View,
  TouchableWithoutFeedback,
  TouchableNativeFeedback,
  Platform
} from 'react-native'
import AppText from '../AppText'
import useStyles from './useStyles'
import { AppButtonProps } from './types'
import CONSTS from './consts'

function AppButton({
  title = CONSTS.BUTTON_TITLE,
  fontSize,
  height,
  backgroundColor,
  borderRadius,
  style,
  onPress
}: AppButtonProps) {
  // const handlePress = useCallback((e: GestureResponderEvent) => { // not recommend!
  //   typeof onPress === 'function' && onPress(e)
  // }, [])
  const styles = useStyles({ fontSize, height, backgroundColor, borderRadius })
  const button = (
    <View style={styles.button}>
      <AppText style={styles.text}>{title}</AppText>
    </View>
  )
  const isAndroid = Platform.OS === 'android'
  return (
    <View style={[styles.buttonBox, style]}>
      {isAndroid ? (
        <TouchableNativeFeedback onPress={onPress}>
          {button}
        </TouchableNativeFeedback>
      ) : (
        <TouchableWithoutFeedback onPress={onPress}>
          {button}
        </TouchableWithoutFeedback>
      )}
    </View>
  )
}

export default AppButton
