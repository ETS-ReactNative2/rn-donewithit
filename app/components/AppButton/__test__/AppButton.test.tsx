import React from 'react'
import renderer from 'react-test-renderer'
import { StyleSheet } from 'react-native'
import AppButton from '../index'

const styles = StyleSheet.create({
  button: {
    marginHorizontal: 10,
    marginBottom: 10
  }
})

const props = {
  style: styles.button,
  title: 'Click'
}

describe('AppButton', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<AppButton {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
