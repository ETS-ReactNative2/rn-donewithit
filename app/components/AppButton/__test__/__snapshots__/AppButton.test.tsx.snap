// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppButton renders correctly 1`] = `
<View
  style={
    Array [
      Object {},
      Object {
        "marginBottom": 10,
        "marginHorizontal": 10,
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "tomato",
        "borderRadius": 25,
        "height": 50,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "Courier",
            "fontSize": 13,
          },
          Object {
            "color": "#fff",
            "fontSize": 18,
            "fontWeight": "bold",
            "textTransform": "uppercase",
          },
        ]
      }
    >
      Click
    </Text>
  </View>
</View>
`;
